### DISCUSSION TITLE
set header on HttpError?

### DISCUSSION THREAD

**Discussion by kdwarn** [oxidecomputer/dropshot#739] [Q&A]
Hello and thanks for building this framework!

Is there a way to set a header on an `HttpError` from an endpoint? I don't see a way to do this, but I'm not sure if I'm just missing something. It would be useful at least in the case of returning a 401 UNAUTHORIZED response, which is supposed to include the [WWW-AUTHENTICATE](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) header. I think a workaround could be to create and return an `Err(Response)` instead of `Err(HttpError)`, but there is a lot of convenience and consistency with using `HttpError`. I've dug a bit into the documentation and code and I'm not sure exactly where the `HttpError` is turned into a `Response` (the documentation for `HttpError` says "When these bubble up to the top of the request handling stack (which is most of the time that they’re generated), these are turned into an HTTP response"), but I think if a `headers` field (a `Vec` of [`&HeaderMap<HeaderValue>`](https://docs.rs/http/latest/http/header/struct.HeaderMap.html)) were added to `HttpError` it could be handled/set there.

**Comment by davepacheco** [comment]
I don't think there's currently a way to do this.  Yes, we've wanted the same thing for `www-authenticate`.

We added support for both structured and unstructured headers on successful responses in #283.  We may want to do something similar here -- i.e., maybe the OpenAPI spec for the 401 response should say that it includes this header?  I'm not sure.

I believe this is where the error is turned into a response:
https://github.com/oxidecomputer/dropshot/blob/6151bb45f2379d96c974dc6568ef5dcb1317cb15/dropshot/src/server.rs#L827
which calls:
https://github.com/oxidecomputer/dropshot/blob/6151bb45f2379d96c974dc6568ef5dcb1317cb15/dropshot/src/error.rs#L235-L267

It'll take some thought to figure out how to best express this in a compatible way.
